fl4health/server/base_server.py

```diff
@@ -350,17 +365,198 @@ def __init__(
+
+            # Evaluate model on a sample of available clients
+            res_fed = self.evaluate_round(server_round=self.current_round, timeout=timeout)
+            if res_fed:
+                loss_fed, evaluate_metrics_fed, _ = res_fed
+                if loss_fed:
+                    self.history.add_loss_distributed(server_round=self.current_round, loss=loss_fed)
+                    self.history.add_metrics_distributed(server_round=self.current_round, metrics=evaluate_metrics_fed)
+
+            self.current_round += 1
+
+            # Save checkpoint after training and testing
+            self._hydrate_model_for_checkpointing()
+            self.save_checkpoint()
+
+        # Bookkeeping
+        end_time = timeit.default_timer()
+        elapsed_time = end_time - start_time
+        log(INFO, "FL finished in %s", elapsed_time)
+        return self.history, elapsed_time
+
+    def save_checkpoint(self) -> None:
+        """
+        Save server checkpoint consisting of model, history, server round, metrics reporter and server name.
+            This method can be overridden to add any necessary state to the checkpoint.
+        """
+
+        assert self.per_round_checkpointer is not None
+
+        ckpt = {
+            "model": self.server_model,
+            "history": self.history,
+            "current_round": self.current_round,
+            "metrics_reporter": self.metrics_reporter,
+            "server_name": self.server_name,
+        }
+
+        self.per_round_checkpointer.save_checkpoint(ckpt)
+
+        log(INFO, f"Saving server state to checkpoint at {self.per_round_checkpointer.checkpoint_path}")
+
+    def load_checkpoint(self) -> None:
+        """
+        Load server checkpoint consisting of model, history, server name, current round and metrics reporter.
+            The method can be overridden to add any necessary state when loading the checkpoint.
+        """
+        assert self.per_round_checkpointer is not None and self.per_round_checkpointer.checkpoint_exists()
+
+        ckpt = self.per_round_checkpointer.load_checkpoint()
+
+        assert "model" in ckpt and isinstance(ckpt["model"], nn.Module)
+        assert "server_name" in ckpt and isinstance(ckpt["server_name"], str)
+        assert "current_round" in ckpt and isinstance(ckpt["current_round"], int)
+        assert "metrics_reporter" in ckpt and isinstance(ckpt["metrics_reporter"], MetricsReporter)
+        assert "history" in ckpt and isinstance(ckpt["history"], History)
+
+        log(INFO, f"Loading server state from checkpoint at {self.per_round_checkpointer.checkpoint_path}")
+
+        self.current_round = ckpt["current_round"]
+        self.server_name = ckpt["server_name"]
+        self.metrics_reporter = ckpt["metrics_reporter"]
+        self.history = ckpt["history"]
+        self.parameters = get_initial_model_parameters(ckpt["model"])
```

Comment:
I don't think we want to rely on the model architecture to get the parameters, but rather should dump the self.parameters that the server has. The reason for this is situations where we are exchanging more information with the server than just model parameters (FedProx, SCAFFOLD, etc.). This will fail when that is happening. I'm also not sure that this will work properly or the way we want it to when we're doing partial exchange?

Reply:
Absolutely, definitely agree! Just for this PR, it seemed to much too tackle supporting all methods, so this was one of the compromises I made and pushed it to the refactor